File: sdk/reproit-backend-php/test/behavior_vectors_test.php
<?php

declare(strict_types=1);

/*
 * Shared behavioral conformance vectors (sdk/capture-behavior-v1.json).
 *
 * Eleven SDKs hand implement one contract, so a defect otherwise has to be
 * found eleven times. Four instances of one class landed in a single day, and
 * every group here was written against one of them.
 */

require_once __DIR__ . '/../reproit.php';

$vectors = json_decode(
    (string) file_get_contents(__DIR__ . '/../../capture-behavior-v1.json'),
    true,
    512,
    JSON_THROW_ON_ERROR
);

$passed = 0;
$failed = 0;

function check(string $label, bool $condition, string $detail = ''): void
{
    global $passed, $failed;
    if ($condition) {
        $passed++;
        return;
    }
    $failed++;
    fwrite(STDERR, "FAIL {$label}" . ($detail === '' ? '' : ": {$detail}") . "\n");
}

// Constants
check(
    'body bound matches vectors',
    \ReproitBackend\MAX_EXCHANGE_BODY_BYTES === $vectors['constants']['maxExchangeBodyBytes']
);
check(
    'divergence marker matches vectors',
    \ReproitBackend\DIVERGENCE_MARKER === $vectors['constants']['divergenceMarker']
);

// Redaction, type cases
foreach ($vectors['redaction']['typeCases'] as $case) {
    $actual = \ReproitBackend\redact($case['input']);
    check(
        'redaction type ' . json_encode($case['input']),
        json_encode($actual) === json_encode($case['expect']),
        'got ' . json_encode($actual) . ' want ' . json_encode($case['expect'])
    );
}

// Redaction, key folding
foreach ($vectors['redaction']['foldingCases'] as $case) {
    $out = \ReproitBackend\redact([$case['field'] => 'value']);
    $value = $out[$case['field']];
    $redacted = is_array($value) && array_key_exists('$reproit', $value);
    check(
        'folding ' . $case['field'],
        $redacted === $case['secret'],
        $case['secret'] ? 'should be secret' : 'should not be secret'
    );
}

// Redaction, nesting
foreach ($vectors['redaction']['nestingCases'] as $case) {
    $actual = \ReproitBackend\redact($case['input']);
    check(
        'nesting ' . json_encode($case['input']),
        json_encode($actual) === json_encode($case['expect']),
        'got ' . json_encode($actual)
    );
}

// The trigger token vocabulary; iOS and RN both shipped user-action.
$token = $vectors['triggerTokens']['bySdkKind']['backend'];
check('trigger token allowed', in_array($token, $vectors['triggerTokens']['allowed'], true));
$source = (string) file_get_contents(__DIR__ . '/../capture.php');
check('capture.php emits the token', str_contains($source, $token));
foreach ($vectors['triggerTokens']['rejected'] as $bad) {
    check("capture.php does not emit {$bad}", !str_contains($source, "'{$bad}'"));
}

echo "behavior_vectors_test: {$passed} passed, {$failed} failed\n";
exit($failed === 0 ? 0 : 1);

File: sdk/reproit-backend-java/src/test/java/dev/reproit/backend/BehaviorVectorsTest.java
package dev.reproit.backend;

// Shared behavioral conformance vectors (sdk/capture-behavior-v1.json).
//
// Eleven SDKs hand implement one contract, so a defect otherwise has to be
// found eleven times. Four instances of one class landed in a single day, and
// every group here was written against one of them.

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertTrue;

import java.nio.file.Files;
import java.nio.file.Path;
import java.util.List;
import java.util.Map;
import org.junit.jupiter.api.Test;

class BehaviorVectorsTest {

    @SuppressWarnings("unchecked")
    private static Map<String, Object> vectors() throws Exception {
        Path path = Path.of("..", "capture-behavior-v1.json").toAbsolutePath().normalize();
        String raw = Files.readString(path);
        return (Map<String, Object>) Json.parse(raw);
    }

    @SuppressWarnings("unchecked")
    @Test
    void constantsMatchTheSharedVectors() throws Exception {
        Map<String, Object> constants = (Map<String, Object>) vectors().get("constants");
        assertEquals(
                ((Number) constants.get("maxExchangeBodyBytes")).intValue(),
                Exchange.MAX_EXCHANGE_BODY_BYTES);
        assertEquals(constants.get("divergenceMarker"), Replay.DIVERGENCE_MARKER);
    }

    @SuppressWarnings("unchecked")
    @Test
    void redactionTypeVectors() throws Exception {
        Map<String, Object> redaction = (Map<String, Object>) vectors().get("redaction");
        for (Object entry : (List<Object>) redaction.get("typeCases")) {
            Map<String, Object> kase = (Map<String, Object>) entry;
            Object actual = BackendTrace.redact(kase.get("input"));
            assertEquals(
                    Json.canonicalJson(kase.get("expect")),
                    Json.canonicalJson(actual),
                    String.valueOf(kase.get("input")));
        }
    }

    @SuppressWarnings("unchecked")
    @Test
    void redactionKeyFoldingVectors() throws Exception {
        Map<String, Object> redaction = (Map<String, Object>) vectors().get("redaction");
        for (Object entry : (List<Object>) redaction.get("foldingCases")) {
            Map<String, Object> kase = (Map<String, Object>) entry;
            String field = (String) kase.get("field");
            Object out = BackendTrace.redact(Map.of(field, "value"));
            Object value = ((Map<String, Object>) out).get(field);
            boolean redacted =
                    value instanceof Map<?, ?> map && map.containsKey("$reproit");
            assertEquals(kase.get("secret"), redacted, field);
        }
    }

    @SuppressWarnings("unchecked")
    @Test
    void triggerTokenIsInTheProtocolVocabulary() throws Exception {
        Map<String, Object> tokens = (Map<String, Object>) vectors().get("triggerTokens");
        Map<String, Object> bySdkKind = (Map<String, Object>) tokens.get("bySdkKind");
        String token = (String) bySdkKind.get("backend");
        assertTrue(((List<Object>) tokens.get("allowed")).contains(token));

        String source = Files.readString(
                Path.of("src/main/java/dev/reproit/backend/Capture.java"));
        assertTrue(source.contains(token), "Capture.java must emit " + token);
        for (Object bad : (List<Object>) tokens.get("rejected")) {
            assertFalse(
                    source.contains("\"" + bad + "\""),
                    "Capture.java must not emit " + bad);
        }
    }
}
